allow using xpath expressions instead of a property name

nickpell opened the pull request:
https://trello.com/c/TQknZitv/169-case-list-detail-allow-formats-on-calculations-release-2

<img width="337" alt="screen shot 2018-01-10 at 2 21 16 pm" src="https://user-images.githubusercontent.com/1757035/34791082-ddc016d2-f611-11e7-8de8-6d71188b7caf.png">

@orangejenny @ctsims cc: @emord

Commit:
allow using xpath expressions instead of a property name

emord commented on corehq/toggles.py:
```diff
@@ -1453,3 +1453,10 @@ def _commtrackify(domain_name, toggle_is_enabled):
+USE_XPATH_EXPRESSION = StaticToggle(
+    'use_xpath_expression',
+    'Enikshay: Allow specifying an xpath expression instead of a case property in the case list.',
+    TAG_CUSTOM,
```
Any reason that this shouldn't be released more widely for other projects?

emord reviewed (commented):
I think if this is truly custom and only for enikshay you could enable `relevant_environments` on the toggle so it's only available on enikshay's cluster.

emord commented:
Also would this be a better fit for "feature preview" in the future? I think we've had some app level feature flags there before, but I don't know if that's being deprecated

ctsims commented:
I'm signed off on this being a solutions level feature.

@sheelio can you confirm whether it would need any changes to make muster there from your perspective?

sheelio commented:
I'm good with this being a Solutions level feature flag.  I know there are a few other projects that have needed this (i.e. L10K).  

Nick, can I type any expression that would also work in the calculated format option?

sheelio commented:
(We also want to have some basic documentation if we're adding this as a solutions level feature flag).

nickpell commented:
> can I type any expression that would also work in the calculated format option?

That's correct

sheelio commented:
Chatted with Nick.  Couple of changes:
1. Move adding this to a split button (Add Calculated Property in the Add Property button)
2. Add minor text saying the row indicating "Calculated Property" (basically where the use xpath checkbox is right now).  
3. Move to a Solutions feature flag.

orangejenny commented:
@sheelio Would you have reservations about making this *the* way users add calculations, totally replacing the "Calculate" option under Format rather than feature flagging it?

orangejenny commented on corehq/toggles.py:
```diff
@@ -1453,3 +1453,10 @@ def _commtrackify(domain_name, toggle_is_enabled):
     namespaces=[NAMESPACE_OTHER],
 )
+
+USE_XPATH_EXPRESSION = StaticToggle(
```
More descriptive name would be nice - maybe something that includes "case list".

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/case_list_properties.html:
```diff
@@ -45,6 +45,12 @@
                     <td class="col-sm-3" data-bind="css: {'has-error': showWarning}">
                         <div data-bind="html: hqImport('app_manager/js/details/screen_config').CC_DETAIL_SCREEN.getFieldHtml(field.val()), visible: !field.edit"></div>
                         <div data-bind="jqueryElement: field.ui, visible: field.edit"></div>
+                        {% if request|toggle_enabled:'USE_XPATH_EXPRESSION' %}
```
@nickpell Calculations are behind an [add-on](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/add_ons.py#L81-L85), this should also check that's on - it'd be weird to have this show up for an app that has calculations turned off.

sheelio commented:
@orangejenny 

Nick and I chatted about replacing the existing calculate feature but that would require a migration or some work to handle the old-style calculates.  I added a Trello card to track that on the SaaS roadmap.
https://trello.com/c/tjuJN2Qi/17-case-list-remove-calculated-format-option-and-replace-with-new-calculated-property

Commit:
move to Solutions level feature flag

nickpell commented:
@sheelio @orangejenny Just to clarify, we want this behind its own feature flag and the calculations add-on, right?

sheelio commented:
I believe so.

orangejenny commented:
@nickpell 

I played around with this locally and ended up with a suite.xml like this

```<field>
        <header>
          <text>
            <locale id="m1.case_long.case_rating * 100_7.header"/>
          </text>
        </header>
        <template>
          <text>
            <xpath function="rating * 100"/>
          </text>
        </template>
      </field>
```

And lines in app_strings like this: `m1.case_long.case_rating * 100_7.header=Rating * 100`

Having ids with spaces & special characters seems liable to cause problems - is a calculation with an equals sign going to break translations?

orangejenny commented:
@sheelio / @ctsims Nick and I talked some more about LOE of doing a migration and found existing code that does something similar. LOE is likely only 1-2 days to do it now, is it reasonable to do that as part of this work?

sheelio commented:
I think its going to improve support and documentation longer-term if we replace the existing calculate format option now, so I'd be in favour of that. 

One weird bit with the calculate format option is that its tied to a specific case property.  I.e. if I setup a case property called cool_date and set the format to "Calculate", I can write a calculation like "format-date(., "%Y%M") where the . refers to cool_date.  I'm not sure this new feature supports that old behaviour.  So it might be hard to migrate all case list/detail configs.

orangejenny commented:
@sheelio Good point, I forgot about the `.` behavior. I think it's just a matter of calling [interpolate_xpath](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/xpath.py#L72) on the expression, but it is an extra wrinkle. @nickpell what do you think?

Commit:
UI improvements; handle turning off the add-on cleanly

Commit:
use label-info

Commit:
Xpath Expression -> Calculated Property

Commit:
use clearer toggle name

Commit:
remove unused var

nickpell commented:
Updated UI:
<img width="982" alt="screen shot 2018-01-11 at 3 36 14 pm" src="https://user-images.githubusercontent.com/1757035/34846152-3f52de9c-f6e5-11e7-913a-ba01288e8317.png">

orangejenny commented:
UX thoughts: I like the split button approach. Can you try either plain text or `class="text-muted"` instead of a label? I think the light blue is more attention-grabbing than is needed.

Commit:
use label-default

Commit:
add lazy migration

Commit:
kill format Calculate

Commit:
fix locale

Commit:
get xpath correctly

nickpell commented:
I've fixed the locale issue and added a lazy migration to convert the Calculate format into Calculated Properties (handling xpath interpolation).

Updated UI:
<img width="970" alt="screen shot 2018-01-15 at 7 53 30 pm" src="https://user-images.githubusercontent.com/1757035/34967010-d05dfc1e-fa2d-11e7-9b50-33d8ff272661.png">

@orangejenny @sheelio

nickpell commented:
Another thought I had is that we should not feature flag this, because we need to support projects that are currently using the Calculate format.  Calculated Property would still be hidden behind the add on.

sheelio commented:
@nickpell 

If we are removing the "Calculate Format" option, we should not put this behind a feature flag, and just leave this behind the calculate add-on.

sheelio commented:
Also, how do we put the field into the suite (i.e. the ID?).  I think we might have a way to convert a calculation into a hash that the "Calculate ID Mapping" thing does.

nickpell commented:
> Also, how do we put the field into the suite (i.e. the ID?).

The locale takes the format ```case_calculated_property_x``` where ```x``` is an integer assigned at application build time.

Commit:
remove flag

Commit:
update locales in tests

Commit:
support  in Calculated Property

Commit:
fix test

Commit:
remove toggle check

Commit:
remove redundant if statement

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -2087,6 +2088,14 @@ def wrap(cls, data):
 
+        # Lazy migration: xpath expressions from format to first-class property
+        if data.get('format') == 'calculate':
+            property_xpath = PropertyXpathGenerator(None, None, None, super(DetailColumn, cls).wrap(data)).xpath
```
Is it legit/allowed to used named params here? This is pretty unreadable.

nickpell replied:
Agree this is pretty messy.

We could have this [__init__](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/detail_screen.py#L73) take keyword arguments instead of positional arguments.

Another option is to separate out https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/detail_screen.py#L563 into a helper function that takes ```column``` as an argument.

nickpell replied:
Thoughts?

orangejenny replied:
I think either would be fine and would be more readable. Also this definitely does not need to block this PR.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/add_property_button.html:
```diff
@@ -11,13 +11,28 @@
+          <li data-bind="click: addXpathExpression"><a>{% trans "Calculated Property" %}</a></li>
+          <!-- /ko -->
+      </ul>
+    {% else %}
```
Nit: it'd be DRYer to do `{% if request|toggle_enabled:'GRAPH_CREATION' or add_ons.calc_xpath %}` above instead of having the `else` block that repeats the caret and other markup. It's not terrible now, but it'll get more awkward when we add another optional item to this dropdown.

orangejenny commented on corehq/apps/app_manager/tests/data/suite/suite-case-detail-instances.xml:
```diff
@@ -62,7 +62,7 @@
       <header>
         <text>
-          <locale id="m0.case_short.case_name_2.header"/>
+          <locale id="m0.case_short.case_calculated_property_2.header"/>
```
Is using the case name in case detail now interpreted as an xpath calculation? How did that happen? Doesn't seem like a problem - as long as the "Calculated Property" label doesn't appear - just curious.

nickpell replied:
It comes from this calculate column:
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/tests/data/suite/app_case_detail_instances.json#L589

orangejenny replied:
Oh, it's not all case names then, cool.

orangejenny commented on corehq/apps/app_manager/suite_xml/sections/details.py:
```diff
@@ -534,10 +534,14 @@ def _get_column_context(self, column):
             else self.app.build_profiles[self.build_profile_id].langs[0]
-        context = {
-            "xpath_function": escape(get_column_generator(
+        if column.useXpathExpression and True:
```
`and True`?

nickpell replied:
whoops: ce839c4

Commit:
fix bool

orangejenny reviewed (approved):
Thank you for all of the updates - this could have been a feature that introduced a bit of tech debt and product debt, but it's in a good place now.

None of the remaining feedback is blocking.

orangejenny reviewed (changes requested):
Sorry, though of one more thing: since apps get lazily migrated on `wrap`, can't you get rid of the old calculate option in the format dropdown altogether, along with the other code in screen_config.js that handles legacy calculations?

Commit:
fix RemoteRequestSuiteTest

Commit:
remove Calculate option from UI

Commit:
lint

nickpell commented:
> Sorry, though of one more thing: since apps get lazily migrated on wrap, can't you get rid of the old calculate option in the format dropdown altogether, along with the other code in screen_config.js that handles legacy calculations?

Thought I had done this but I guess I missed it.  Updated.

@orangejenny
